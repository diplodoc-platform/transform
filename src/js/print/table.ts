const Selector = {
    WRAPPED_TABLE: '.pdf .yfm-table-container table',
    TABLE: '.pdf table',
};
const Padding = {
    BOTTOM: 20,
};

function resizeElement(element: HTMLElement) {
    const availableWidth = element.parentElement?.offsetWidth;
    const contentWidth = (element.firstElementChild as HTMLElement).offsetWidth;

    if (!availableWidth) {
        return;
    }

    const needScale = contentWidth > availableWidth;

    if (needScale) {
        const scale = availableWidth / contentWidth;

        element.style.transform = `scale(${scale})`;
    }

    element.parentElement.style.height = `${
        element.getBoundingClientRect().height + Padding.BOTTOM
    }px`;
}

function resizeElements() {
    document.querySelectorAll(Selector.WRAPPED_TABLE).forEach((element) => {
        resizeElement(element as HTMLElement);
    });
}

function wrapTables() {
    const tables = document.querySelectorAll(Selector.TABLE);

    for (let i = 0; i < tables.length; i++) {
        const table = tables[i];

        const parent = table.parentNode;

        if (!parent) {
            continue;
        }

        const wrapper = document.createElement('div');
        parent.insertBefore(wrapper, table);
        wrapper.appendChild(table);
        wrapper.classList.add('yfm-table-container');
    }
}

if (typeof document !== 'undefined') {
    window.addEventListener('load', () => {
        wrapTables();
        resizeElements();
    });
}

export {};
