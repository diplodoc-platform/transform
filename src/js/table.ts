import debounce from 'debounce';

const Selector = {
    TABLE: '.yfm-table-container table',
};
const Padding = {
    BOTTOM: 20,
};

function resizeElement(element: HTMLElement) {
    const availableWidth = element.parentElement?.offsetWidth;
    const contentWidth = (element.firstElementChild as HTMLElement).offsetWidth;

    if (!availableWidth) {
        return;
    }

    const needScale = contentWidth > availableWidth;

    if (needScale) {
        const scale = availableWidth / contentWidth;

        element.style.transform = `scale(${scale})`;
    }

    element.parentElement.style.height = `${
        element.getBoundingClientRect().height + Padding.BOTTOM
    }px`;
}

function resizeElements() {
    document.querySelectorAll(Selector.TABLE).forEach((element) => {
        resizeElement(element as HTMLElement);
    });
}

if (typeof document !== 'undefined') {
    window.addEventListener('resize', debounce(resizeElements, 200));

    window.addEventListener('load', () => {
        resizeElements();

        const observer = new MutationObserver(resizeElements);

        const rootEl = document.getElementById('root');
        if (rootEl) {
            observer.observe(rootEl, {childList: true, subtree: true});
        }
    });
}

export {};
