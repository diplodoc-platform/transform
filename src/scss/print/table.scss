.yfm {
    .yfm-table-container {
        position: relative;

        & > table {
            overflow: initial;
            position: absolute;
            top: 0;
            left: 0;
            max-width: initial;
            transform-origin: top left;

            thead {
                display: table-row-group;
            }

            th, td {
                page-break-inside: avoid;
                white-space: pre-wrap;
                text-overflow: initial;
            }
        }
    }
}