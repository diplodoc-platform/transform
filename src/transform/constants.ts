import type {Lang} from './typings';

export const DEFAULT_LANG: Lang = 'ru';
