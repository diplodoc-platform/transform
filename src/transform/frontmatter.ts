import {YAMLException, dump, load} from 'js-yaml';
import {dedent} from 'ts-dedent';
import cloneDeepWith from 'lodash/cloneDeepWith';

import {log} from './log';

export type FrontMatter = {
    [key: string]: unknown;
    metadata?: Record<string, unknown>[];
};

const SEP = '---';

/**
 * Temporary workaround to enable parsing YAML metadata from potentially
 * Liquid-aware source files
 * @param content Input string which could contain Liquid-style substitution syntax (which clashes with YAML
 * object syntax)
 * @returns String with `{}` escaped, ready to be parsed with `js-yaml`
 */
const escapeLiquid = (content: string): string =>
    content.replace(/{{/g, '(({{').replace(/}}/g, '}}))');

/**
 * Inverse of a workaround defined above.
 * @see `escapeLiquidSubstitutionSyntax`
 * @param escapedContent Input string with `{}` escaped with backslashes
 * @returns Unescaped string
 */
const unescapeLiquid = (escapedContent: string): string =>
    escapedContent.replace(/\(\({{/g, '{{').replace(/}}\)\)/g, '}}');

const matchMetadata = (fileContent: string) => {
    if (!fileContent.startsWith(SEP)) {
        return null;
    }

    const closeStart = fileContent.indexOf('\n' + SEP, SEP.length);
    const closeEnd = fileContent.indexOf('\n', closeStart + 1);

    if (closeStart === -1) {
        return null;
    }

    return [fileContent.slice(SEP.length, closeStart).trim(), fileContent.slice(closeEnd + 1)];
};

const duplicateKeysCompatibleLoad = (yaml: string, filePath: string | undefined) => {
    try {
        return load(yaml);
    } catch (e) {
        if (e instanceof YAMLException) {
            const duplicateKeysDeprecationWarning = dedent`
                In ${filePath ?? '(unknown)'}: Encountered a YAML parsing exception when processing file metadata: ${e.reason}.
                It's highly possible the input file contains duplicate mapping keys.
                Will retry processing with necessary compatibility flags.
                Please note that this behaviour is DEPRECATED and WILL be removed in a future version
                without further notice, so the build WILL fail when supplied with YAML-incompatible meta.
            `;

            log.warn(duplicateKeysDeprecationWarning);

            return load(yaml, {json: true});
        }

        throw e;
    }
};

export const extractFrontMatter = (
    fileContent: string,
    filePath?: string,
): [FrontMatter, string] => {
    const matches = matchMetadata(fileContent);

    if (matches) {
        const [metadata, strippedContent] = matches;

        return [
            cloneDeepWith(
                duplicateKeysCompatibleLoad(escapeLiquid(metadata), filePath) as FrontMatter,
                (v) => (typeof v === 'string' ? unescapeLiquid(v) : undefined),
            ),
            strippedContent,
        ];
    }

    return [{}, fileContent];
};

export const composeFrontMatter = (frontMatter: FrontMatter, strippedContent: string) => {
    const dumped = dump(frontMatter, {lineWidth: -1}).trim();

    // This empty object check is a bit naive
    // The other option would be to check if all own fields are `undefined`,
    // since we exploit passing in `undefined` to remove a field quite a bit
    if (dumped === '{}') {
        return strippedContent;
    }

    return `${SEP}\n${dumped}\n${SEP}\n${strippedContent}`;
};
