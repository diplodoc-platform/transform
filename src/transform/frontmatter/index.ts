export * from './extract';
export * from './emplace';
export * from './transformValues';
export {countLineAmount} from './common';
