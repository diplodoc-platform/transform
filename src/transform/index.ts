import {bold} from 'chalk';
import attrs from 'markdown-it-attrs';
import Token from 'markdown-it/lib/token';

import {log, LogLevels} from './log';
import makeHighlight from './highlight';
import extractTitle from './title';
import getHeadings from './headings';
import liquid from './liquid';
import sanitizeHtml, {SanitizeOptions} from './sanitize';

import notes from './plugins/notes';
import anchors from './plugins/anchors';
import code from './plugins/code';
import cut from './plugins/cut';
import deflist from './plugins/deflist';
import term from './plugins/term';
import file from './plugins/file';
import imsize from './plugins/imsize';
import meta from './plugins/meta';
import sup from './plugins/sup';
import tabs from './plugins/tabs';
import video from './plugins/video';
import monospace from './plugins/monospace';
import yfmTable from './plugins/table';
import {initMd} from './md';
import {MarkdownItPluginCb} from './plugins/typings';
import type {HighlightLangMap, Heading} from './typings';

interface OutputType {
    result: {
        html: string;
        title?: string;
        headings: Heading[];
        assets?: unknown[];
        meta?: object;
    };
    logs: Record<LogLevels, string[]>;
}
interface OptionsType {
    vars?: Record<string, string>;
    path?: string;
    extractTitle?: boolean;
    needTitle?: boolean;
    allowHTML?: boolean;
    linkify?: boolean;
    breaks?: boolean;
    conditionsInCode?: boolean;
    disableLiquid?: boolean;
    leftDelimiter?: string;
    rightDelimiter?: string;
    isLiquided?: boolean;
    needToSanitizeHtml?: boolean;
    sanitizeOptions?: SanitizeOptions;
    needFlatListHeadings?: boolean;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    plugins?: MarkdownItPluginCb<any>[];
    highlightLangs?: HighlightLangMap;
    root?: string;
    [x: string]: unknown;
}

function transform(originInput: string, opts: OptionsType = {}): OutputType {
    const {
        vars = {},
        path,
        extractTitle: extractTitleOption,
        needTitle,
        allowHTML = false,
        linkify = false,
        breaks = true,
        conditionsInCode = false,
        needToSanitizeHtml = false,
        sanitizeOptions,
        needFlatListHeadings = false,
        disableLiquid = false,
        leftDelimiter = '{',
        rightDelimiter = '}',
        isLiquided = false,
        plugins = [
            meta,
            deflist,
            cut,
            notes,
            anchors,
            tabs,
            code,
            sup,
            video,
            monospace,
            yfmTable,
            file,
            imsize,
            term,
        ],
        highlightLangs = {},
        ...customOptions
    } = opts;

    const pluginOptions = {
        ...customOptions,
        vars,
        path,
        extractTitle: extractTitleOption,
        disableLiquid,
        log,
    };

    const input =
        disableLiquid || isLiquided
            ? originInput
            : liquid(originInput, vars, path, {conditionsInCode});

    const highlight = makeHighlight(highlightLangs);
    const md = initMd({html: allowHTML, linkify, highlight, breaks});
    // Need for ids of headers
    md.use(attrs, {leftDelimiter, rightDelimiter});

    plugins.forEach((plugin) => md.use(plugin, pluginOptions));

    try {
        let title;
        let tokens;
        let titleTokens;
        const env = {} as {[key: string]: Token[] | unknown};

        tokens = md.parse(input, env);

        if (extractTitleOption) {
            ({title, tokens, titleTokens} = extractTitle(tokens));

            // title tokens include other tokens that need to be transformed
            if (titleTokens.length > 1) {
                title = md.renderer.render(titleTokens, md.options, env);
            }
        }
        if (needTitle) {
            ({title} = extractTitle(tokens));
        }

        const headings = getHeadings(tokens, needFlatListHeadings);

        // add all term template tokens to the end of the html
        const termTokens = (env.termTokens as Token[]) || [];
        let html = md.renderer.render([...tokens, ...termTokens], md.options, env);
        if (needToSanitizeHtml) {
            html = sanitizeHtml(html, sanitizeOptions);
        }

        const assets = md.assets;
        const meta = md.meta;

        return {
            result: {html, title, headings, assets, meta},
            logs: log.get(),
        };
    } catch (err) {
        log.error(`Error occurred${path ? ` in ${bold(path)}` : ''}`);
        throw err;
    }
}

export = transform;

// eslint-disable-next-line @typescript-eslint/no-namespace -- backward compatibility
namespace transform {
    export type Options = OptionsType;
    export type Output = OutputType;
}
