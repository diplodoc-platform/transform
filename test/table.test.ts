import transform from '../src/transform';
import table from '../src/transform/plugins/table';

const transformYfm = (text: string) => {
    const {
        result: {html},
    } = transform(text, {
        plugins: [table],
    });
    return html;
};

describe('Table plugin', () => {
    it('should render simple inline table', () => {
        expect(transformYfm('#|\n||1|2||\n||3|4||\n|#')).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>2</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>3</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>4</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should render simple table', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '||Cell in column 1, row 1\n' +
                    '|Cell in column 2, row 1||\n' +
                    '||Cell in column 1, row 2\n' +
                    '|Cell in column 2, row 2||\n' +
                    '||Cell in column 1, row 3\n' +
                    '|Cell in column 2, row 3||\n' +
                    '|#',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Cell in column 1, row 1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>Cell in column 2, row 1</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Cell in column 1, row 2</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>Cell in column 2, row 2</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Cell in column 1, row 3</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>Cell in column 2, row 3</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should render table between paragraphs', () => {
        expect(
            transformYfm(
                'Text\n' +
                    '\n' +
                    '#|\n' +
                    '||Cell in column 1, row 1\n' +
                    '|Cell in column 2, row 1||\n' +
                    '||Cell in column 1, row 2\n' +
                    '|Cell in column 2, row 2||\n' +
                    '|#\n' +
                    '\n' +
                    'Text',
            ),
        ).toBe(
            '<p>Text</p>\n' +
                '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Cell in column 1, row 1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>Cell in column 2, row 1</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Cell in column 1, row 2</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>Cell in column 2, row 2</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n' +
                '<p>Text</p>\n',
        );
    });
    it('should render table with inline elements', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '||**bold**\n' +
                    '|_italic_||\n' +
                    '||_**bold and italic**_\n' +
                    '|`code snippet`||\n' +
                    '|| ![alt text](image.png)\n' +
                    '| [link to README.md](README.md)||\n' +
                    '|#',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p><strong>bold</strong></p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p><em>italic</em></p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p><em><strong>bold and italic</strong></em></p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p><code>code snippet</code></p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p><img src="image.png" alt="alt text" /></p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p><a href="README.md">link to README.md</a></p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should render table with block elements', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '||Text with\n' +
                    'new line\n' +
                    '|- Element 1\n' +
                    '- Element 2\n' +
                    '- Element 3\n' +
                    '- Element 4||\n' +
                    '||\n' +
                    '- Element 1\n' +
                    '   - Element A\n' +
                    '   - Element B\n' +
                    '- Element 2\n' +
                    'test||\n' +
                    '|| test\n' +
                    '# h1\n' +
                    'test\n' +
                    '|test||\n' +
                    '|#\n',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Text with<br />\n' +
                'new line</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<ul>\n' +
                '<li>Element 1</li>\n' +
                '<li>Element 2</li>\n' +
                '<li>Element 3</li>\n' +
                '<li>Element 4</li>\n' +
                '</ul>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<ul>\n' +
                '<li>Element 1\n' +
                '<ul>\n' +
                '<li>Element A</li>\n' +
                '<li>Element B</li>\n' +
                '</ul>\n' +
                '</li>\n' +
                '<li>Element 2<br />\n' +
                'test</li>\n' +
                '</ul>\n' +
                '</td>\n' +
                '<td></td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>test</p>\n' +
                '<h1>h1</h1>\n' +
                '<p>test</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>test</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should render two tables', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '|| 1\n' +
                    '| 2||\n' +
                    '|| 3\n' +
                    '| 4||\n' +
                    '|#\n' +
                    '\n' +
                    '#|\n' +
                    '|| 5\n' +
                    '| 6||\n' +
                    '|| 7\n' +
                    '| 8||\n' +
                    '|#',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>2</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>3</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>4</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n' +
                '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>5</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>6</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>7</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>8</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should not render table inside code', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '|| 1\n' +
                    '|\n' +
                    '```\n' +
                    '#|\n' +
                    '|| 4\n' +
                    '| 5||\n' +
                    '\n' +
                    '|| 6\n' +
                    '| 7||\n' +
                    '|#\n' +
                    '```||\n' +
                    '|| 2\n' +
                    '| 3||\n' +
                    '|#',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<pre><code class="hljs">#|\n' +
                '|| 4\n' +
                '| 5||\n' +
                '\n' +
                '|| 6\n' +
                '| 7||\n' +
                '|#\n' +
                '</code></pre>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>2</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>3</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should not render table without close token', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '||Cell in column 1, row 1\n' +
                    '|Cell in column 2, row 1||\n' +
                    '||Cell in column 1, row 2\n' +
                    '|Cell in column 2, row 2||\n' +
                    '||Cell in column 1, row 3\n' +
                    '|Cell in column 2, row 3||',
            ),
        ).toBe(
            '<p>#|<br />\n' +
                '||Cell in column 1, row 1<br />\n' +
                '|Cell in column 2, row 1||<br />\n' +
                '||Cell in column 1, row 2<br />\n' +
                '|Cell in column 2, row 2||<br />\n' +
                '||Cell in column 1, row 3<br />\n' +
                '|Cell in column 2, row 3||</p>\n',
        );
    });
    it('should not render table without open token', () => {
        expect(
            transformYfm(
                '||Cell in column 1, row 1\n' +
                    '|Cell in column 2, row 1\n' +
                    '||Cell in column 1, row 2\n' +
                    '|Cell in column 2, row 2\n' +
                    '|#\n' +
                    '\n' +
                    'Test',
            ),
        ).toBe(
            '<p>||Cell in column 1, row 1<br />\n' +
                '|Cell in column 2, row 1<br />\n' +
                '||Cell in column 1, row 2<br />\n' +
                '|Cell in column 2, row 2<br />\n' +
                '|#</p>\n' +
                '<p>Test</p>\n',
        );
    });
    it('should not render table if no empty line before table', () => {
        expect(
            transformYfm(
                'Test\n' +
                    '#|\n' +
                    '||Cell in column 1, row 1\n' +
                    '|Cell in column 2, row 1||\n' +
                    '\n' +
                    '||Cell in column 1, row 2\n' +
                    '|Cell in column 2, row 2||\n' +
                    '|#',
            ),
        ).toBe(
            '<p>Test<br />\n' +
                '#|<br />\n' +
                '||Cell in column 1, row 1<br />\n' +
                '|Cell in column 2, row 1||</p>\n' +
                '<p>||Cell in column 1, row 2<br />\n' +
                '|Cell in column 2, row 2||<br />\n' +
                '|#</p>\n',
        );
    });
    it('should render table inside table', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '|| 1\n' +
                    '|\n' +
                    '\n' +
                    'test\n' +
                    '\n' +
                    '#|\n' +
                    '|| 5\n' +
                    '| 6||\n' +
                    '|| 7\n' +
                    '| 8||\n' +
                    '|#\n' +
                    '\n' +
                    'test||\n' +
                    '|| 3\n' +
                    '| 4||\n' +
                    '|#',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>test</p>\n' +
                '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>5</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>6</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>7</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>8</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n' +
                '<p>test</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>3</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>4</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should render table with spaces after new line', () => {
        expect(
            transformYfm(
                '   #|\n' +
                    '  || row 1 col 1\n' +
                    '   | row 1 col 2||\n' +
                    '  || row 2 col 1\n' +
                    ' | row 2 col 2 ||\n' +
                    '    |#',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>row 1 col 1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>row 1 col 2</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>row 2 col 1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>row 2 col 2</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should add empty cells', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '|| row 1 col 1\n' +
                    '| row 1 col 2||\n' +
                    '|| row 2 col 1\n' +
                    '| row 2 col 2||\n' +
                    '|| row 3 col 1\n' +
                    '| row 3 col 2\n' +
                    '| row 3 col 3||\n' +
                    '|#',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>row 1 col 1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>row 1 col 2</p>\n' +
                '</td>\n' +
                '<td></td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>row 2 col 1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>row 2 col 2</p>\n' +
                '</td>\n' +
                '<td></td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>row 3 col 1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>row 3 col 2</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>row 3 col 3</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should render table inside quote on each line', () => {
        expect(
            transformYfm(
                '> #|\n' +
                    '> ||Cell in column 1, row 1\n' +
                    '> |Cell in column 2, row 1||\n' +
                    '> ||Cell in column 1, row 2\n' +
                    '> |Cell in column 2, row 2||\n' +
                    '> ||Cell in column 1, row 3\n' +
                    '> |Cell in column 2, row 3||\n' +
                    '> |#',
            ),
        ).toBe(
            '<blockquote>\n' +
                '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Cell in column 1, row 1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>Cell in column 2, row 1</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Cell in column 1, row 2</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>Cell in column 2, row 2</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>Cell in column 1, row 3</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>Cell in column 2, row 3</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n' +
                '</blockquote>\n',
        );
    });
    it('should correct render escaped cells', () => {
        expect(transformYfm('#|\n||1 \\| 1.1|2||\n||3|4||\n|#')).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>1 | 1.1</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>2</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>3</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>4</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
    it('should correct render escaped rows', () => {
        expect(
            transformYfm(
                '#|\n' +
                    '|| cell text \\|| cell text ||\n' +
                    '|| cell text | cell text ||\n' +
                    '|#',
            ),
        ).toBe(
            '<table>\n' +
                '<tbody>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>cell text |</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>cell text</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '<tr>\n' +
                '<td>\n' +
                '<p>cell text</p>\n' +
                '</td>\n' +
                '<td>\n' +
                '<p>cell text</p>\n' +
                '</td>\n' +
                '</tr>\n' +
                '</tbody>\n' +
                '</table>\n',
        );
    });
});
