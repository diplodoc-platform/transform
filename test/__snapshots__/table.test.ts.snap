// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table plugin parses table with liquid variables inside 1`] = `
"<table>
<tbody>
<tr>
<td>
<p>{{variable | length}} column 1, row 1</p>
</td>
<td>
<p>Cell in column 2, row 1</p>
</td>
</tr>
<tr>
<td>
<p>Cell in column 1, row 2</p>
</td>
<td>
<p>Cell in column 2, row 2</p>
</td>
</tr>
</tbody>
</table>
"
`;
